### PR TITLE
Storage/STG74 Bearer Challenge

### PR DESCRIPTION
**Pull request by microzchang**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:21]:
```suggestion
        std::shared_ptr<const Azure::Core::Credentials::TokenCredential> credential,
```

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:21]: keep consistent with existing code

File: sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp:40]: Is this thread safe?

File: sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp:34]: How is this different from the test case above? You provided tenant ID. How does the new policy make things different?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:23]: Why this parameter has a default value?
It's always set with `options.EnableTenantDiscovery`.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/rest_client.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-queues/inc/azure/storage/queues/rest_client.hpp:29]: Don't forget to update CHANGELOG.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:39]: why do we need `mutable` here?

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp
Comment (microzchang) [reply sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp:34]: This tenantId is in credential,  and will not pass to StorageBearerTokenAuthenticationPolicy. So when enableTenantDiscovery, it will trigger bearer challenge.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:21]: For example, which existing code?
I keeped consistent with [BearerTokenAuthenticationPolicy](https://github.com/Azure/azure-sdk-for-cpp/blob/b33e8a9fd43a39762ac4e648336cd110c5fc58ab/sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp#L548) in current code.

File: sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp:40]: Yeah, it's not thread safe and I add a lock to ensure this.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:39]: Because this field may be changed in a [const method](https://github.com/microzchang/azure-sdk-for-cpp/blob/6727a4381ae5e11e4e243a6f9911329dbb696405/sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp#L40)

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:23]: Fixed.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:21]: Fixed

File: sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp:34]: what if `enableTenantDiscovery` is not enable? `m_TenantId` is always initialized as empty. 

https://github.com/microzchang/azure-sdk-for-cpp/blob/6727a4381ae5e11e4e243a6f9911329dbb696405/sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp#L22

is it a problem if this field at this line is empty?

Comment (microzchang) [reply sdk/storage/azure-storage-blobs/test/ut/bearer_token_test.cpp:34]: No, there is no problem when m_TenantId is empty. The tenantId is always empty in [the old logic](https://github.com/Azure/azure-sdk-for-cpp/blob/b33e8a9fd43a39762ac4e648336cd110c5fc58ab/sdk/storage/azure-storage-blobs/src/blob_client.cpp#L89C37-L89C37) and [ClientSecretCredential](https://github.com/Azure/azure-sdk-for-cpp/blob/b33e8a9fd43a39762ac4e648336cd110c5fc58ab/sdk/identity/azure-identity/src/client_secret_credential.cpp#L73) can [resolve tenantId from both credential and tokenRequestContext](https://github.com/Azure/azure-sdk-for-cpp/blob/b33e8a9fd43a39762ac4e648336cd110c5fc58ab/sdk/identity/azure-identity/src/tenant_id_resolver.cpp#L29).

File: sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp:42]: This is not the correct usage of mutex.

If you protect a variable with a mutex, you'll need to hold the mutex when reading and writing to the variable (Only exception is in constructor and destructor).

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:45]: still thread unsafe

File: sdk/storage/azure-storage-queues/CHANGELOG.md
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-queues/CHANGELOG.md:9]:
```suggestion
    - A new property is now available on `QueueClientOptions` called `EnableTenantDiscovery`. If set to `true`, the client will attempt an initial unauthorized request to the service to prompt a challenge containing the tenantId hint.
```

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:41]:
```suggestion
    std::vector<std::string> m_scopes;
    mutable std::string m_tenantId;
    mutable std::mutex m_tenantIdMutex;
    bool m_enableTenantDiscovery;
```

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:40]: You can do better than the plain `std::mutex`  - use `std::shared_timed_mutex` (unlike `std::shared_mutex`, it is available in C++14), and then you can do `std::shared_lock` for reading and `std::unique_lock` for writing.
For examples, one of the places we're doing this, is in core logging functionality: https://github.com/Azure/azure-sdk-for-cpp/blob/0b59e5a94c3551cf550196db1ee7175972e69d10/sdk/core/azure-core/src/logger.cpp

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp]: Fixed, thanks for review.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp]: Thanks for your suggestion. I will use `std::shared_timed_mutex` and improve this code.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:8]: Technically, `std::unique_lock` is defined in `<mutex>`, and including `<shared_mutex>` alone does not have to make it available, so it is better to explicitly include both headers.
```suggestion
#include <mutex>
#include <shared_mutex>
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp
Comment (antkmsft) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:167]: If you want, I think it is ok to not have this option (or delay having it) - effectively, `"EnableTenantDiscovery"` will be `"true"` when the credential is going to have something in the credential options'  `AdditionallyAllowedTenants` vector - additional tenant IDs, or a `"*"` value.
If it is empty (and it is empty by default), then the policy will effectively work like `BearerTokenAuthenticationPolicy` does today - in either case, even if you parse the TenantId from the challenge and set that, the credential will not let you to authorize via another TenantId, unless it is in its `options.AdditionallyAllowedTenants`.

Comment (antkmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:167]: Otherwise, it think it will be more consistent to return `false` from `AuthorizeRequestOnChallenge()` right away if `EnableTenantDiscovery` is `false`. But up to you.

File: sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp:19]: Does this condition work as you intended? Did you mean `&&`?
* If tenant discovery is disabled, it is still going to override tenantId with an empty one. In which case it is going to get the challenge response, and your `AuthorizeRequestOnChallenge()` method will be called.
* Unless authentication was successful, `AuthorizeRequestOnChallenge()` will be called when we get a challenge response, even if you did not override tenantId in `AuthorizeAndSendRequest()`, so it is going to set `m_tenantId`, and unless it returns `false`, another attempt will be made to get a token, with the new tenantId.
* Next time it comes to authorization, your `m_tenantId` is not empty, so you ignore `m_enableTenantDiscovery`.

**Review by antkmsft — COMMENTED** [review]
I think it will be clearer how the policy works, and if it works as intended, if you add tests, despite that it is a good practice. For inspiration, you can look into https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/keyvault/azure-security-keyvault-secrets/test/ut/challenge_based_authentication_policy_test.cpp - you can see if `EnableTenantDiscovery` works as intended, and so on.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:46]: fixed

File: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:47]: Is `SafeTenantId` or `StorageBearerTokenAuthenticationPolicy` required to be copy-assignable?

Comment (microzchang) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp:47]: No, I think clone() method is enough.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp
Comment (microzchang) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:167]: Yeah, I think it's make sense. I've add thelogic to return `false `from `AuthorizeRequestOnChallenge()` right away if `EnableTenantDiscovery` is `false`.

File: sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp
Comment (microzchang) [reply sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp:19]: Yeah, this condition is generally correct. 

And the reason why I add `EnableTenantDiscovery` is that the old logic will attmpt to get a token from AAD before real request, so if tenantId is not set in `ClientSecretCredential`, a `404` error from AAD will be received, and can't trigger any bearer challenge. 

And when `EnableTenantDiscovery` is true, it will directly send request without token, and received a 401 bearer challange. With the tenantId from challange, attempt will be made to get a token. So even if the tenantId is empty in `ClientSecretCredential`, it can get the correct tenantId from bearer challenge.

**Review by Jinming-Hu — APPROVED** [review]
This looks good to me.


@antkmsft Can you take another look at this PR? We'd like to have your approval before we hit the merge button.

**Review by antkmsft — APPROVED** [review]

**Comment by antkmsft** [comment]
Thank you @Jinming-Hu, looks good to me, but as I said, I think it would've been more beneficial for Storage as well, if you had test coverage for it ensuring that it works just as you intended.

**Comment by microzchang** [comment]
> Thank you @Jinming-Hu, looks good to me, but as I said, I think it would've been more beneficial for Storage as well, if you had test coverage for it ensuring that it works just as you intended.

Thanks for your suggestion. I've added several scenarios to test the different conditions about `EnableTenantDiscovery` and bearer challenge. And storage bearer challenge policy doesn't support change scope from challallenge.

**Comment by antkmsft** [comment]
Thank you @microzchang!

As for the CI failure to run `/azp` command, I've also seen that a bit earlier, and I ran the pipeline manually. Something has broke recently with the CI, I think we need to contact EngSys team.

**Comment by ahsonkhan** [comment]
> Something has broke recently with the CI, I think we need to contact EngSys team.

Here's the issue tracking that: https://github.com/Azure/azure-sdk-for-cpp/issues/4779

**Comment by microzchang** [comment]
Theanually triggered pipeline has been successful: [pipeline](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2923422&view=results)

**Pull request merged**
